File: js/src/cv.ts
/** CV accession ↔ integer-tail mapping (mirrors the Python `cv` module). */

import {
  ARRAY_CHARGE,
  ARRAY_INTENSITY,
  ARRAY_MZ,
  ARRAY_NON_STANDARD,
  COMP_NUMLIN_ZLIB,
  COMP_NUMPIC_ZLIB,
  COMP_NUMSLOF_ZLIB,
  COMP_ZLIB,
  COMP_ZSTD,
  COMP_BYTE_SHUFFLED_ZSTD,
  COMP_NUMLIN_ZSTD,
  COMP_NUMPIC_ZSTD,
  COMP_NUMSLOF_ZSTD,
  ION_MOBILITY_ARRAY_TAILS,
  TYPE_FLOAT32,
  TYPE_FLOAT64,
  TYPE_INT32,
} from "./format.js";

export {
  ARRAY_CHARGE,
  ARRAY_INTENSITY,
  ARRAY_MZ,
  ARRAY_NON_STANDARD,
  COMP_NUMLIN_ZLIB,
  COMP_NUMPIC_ZLIB,
  COMP_NUMSLOF_ZLIB,
  COMP_ZLIB,
  COMP_ZSTD,
  COMP_BYTE_SHUFFLED_ZSTD,
  COMP_NUMLIN_ZSTD,
  COMP_NUMPIC_ZSTD,
  COMP_NUMSLOF_ZSTD,
  ION_MOBILITY_ARRAY_TAILS,
  TYPE_FLOAT32,
  TYPE_FLOAT64,
  TYPE_INT32,
};

const DEFAULT_PARAM_ONTOLOGY = "MS";
const DEFAULT_UNIT_ONTOLOGY = "UO";

export function accessionTail(accession: string): number {
  return parseInt(accession.split(":")[1]!, 10);
}

export function accessionOntology(accession: string): string {
  return accession.split(":")[0]!;
}

function pad7(tail: number): string {
  return String(tail).padStart(7, "0");
}

export function decodeTail(tail: number, ontology: string = DEFAULT_PARAM_ONTOLOGY): string {
  return `${ontology}:${pad7(tail)}`;
}

/** True when the accession's tail is exactly 7 digits, i.e. survives the
 * 7-digit zero-padded reconstruction used for tail-encoded keys. */
function hasSevenDigitTail(accession: string): boolean {
  const idx = accession.indexOf(":");
  return idx >= 0 && /^\d{7}$/.test(accession.slice(idx + 1));
}

/**
 * Encode an accession as a param-map key: a bare integer tail for `MS:` keys
 * with 7-digit tails, or the full accession string (e.g. `"UO:0000031"`,
 * `"NCIT:C25330"`) otherwise, keeping the key a hashable CBOR-map scalar.
 */
export function encodeParamKey(accession: string): number | string {
  return accessionOntology(accession) === DEFAULT_PARAM_ONTOLOGY && hasSevenDigitTail(accession)
    ? accessionTail(accession)
    : accession;
}

export function decodeParamKey(key: number | string): string {
  return typeof key === "string" ? key : decodeTail(key);
}

/** Encode a unit accession: a bare tail (UO:), `[ontology, tail]` for other
 * ontologies, or the full accession string when the tail is not exactly 7
 * digits (e.g. `"MOD:00046"`, which would not round-trip through padding). */
export function encodeUnit(unitAccession: string): number | [string, number] | string {
  if (!hasSevenDigitTail(unitAccession)) return unitAccession;
  const onto = accessionOntology(unitAccession);
  const tail = accessionTail(unitAccession);
  return onto === DEFAULT_UNIT_ONTOLOGY ? tail : [onto, tail];
}

export function decodeUnitTail(t: number | [string, number] | string): string {
  if (typeof t === "string") return t;
  if (Array.isArray(t)) return `${t[0]}:${pad7(t[1])}`;
  return `${DEFAULT_UNIT_ONTOLOGY}:${pad7(t)}`;
}

File: src/spectrl/cv.py
"""CV accession ↔ integer-tail mapping for the spectrl wire format.

Rules (§3.1):
- Accession tails default to MS: ontology.
- Unit tails default to UO: ontology.
- Any other ontology uses an explicit [ontology_id, tail] pair.

The tail for "MS:1000511" is 1000511; for "UO:0000031" is 31.
"""

from __future__ import annotations

import re

from ._format import ARRAY_CHARGE as ARRAY_CHARGE
from ._format import ARRAY_INTENSITY as ARRAY_INTENSITY
from ._format import ARRAY_MZ as ARRAY_MZ
from ._format import ARRAY_NON_STANDARD as ARRAY_NON_STANDARD
from ._format import ION_MOBILITY_ARRAY_TAILS as _ION_MOBILITY_TAILS
from ._format import TYPE_FLOAT32 as TYPE_FLOAT32
from ._format import TYPE_FLOAT64 as TYPE_FLOAT64
from ._format import TYPE_INT32 as TYPE_INT32

_DEFAULT_PARAM_ONTOLOGY = "MS"
_DEFAULT_UNIT_ONTOLOGY = "UO"
# Largest tail that survives seven-digit zero-padded reconstruction (section 3).
_MAX_NUMERIC_TAIL = 9999999
_ACCESSION_RE = re.compile(r"[A-Za-z][A-Za-z0-9]*:[A-Za-z0-9]+")
_PREFIX_RE = re.compile(r"[A-Za-z][A-Za-z0-9]*")


def accession_tail(accession: str) -> int:
    """Extract the integer tail from an accession string like 'MS:1000511' → 1000511."""
    try:
        return int(accession.split(":", 1)[1])
    except (IndexError, ValueError):
        raise ValueError(
            f"accession {accession!r} has no integer tail; accessions with non-numeric tails "
            "are carried as full strings and cannot be tail-encoded."
        ) from None


def accession_ontology(accession: str) -> str:
    """Extract the ontology prefix from 'MS:1000511' → 'MS'."""
    return accession.split(":")[0]


def encode_unit(unit_accession: str) -> int | list | str:
    """Encode a unit accession: tail int (UO: default), [ontology, tail] for other
    ontologies, or the full accession string when the tail is not exactly 7 digits
    (tail encoding reconstructs with 7-digit zero-padding, so anything else would
    not round-trip, e.g. 'MOD:00046')."""
    onto = accession_ontology(unit_accession)
    tail_str = unit_accession.split(":", 1)[1] if ":" in unit_accession else ""
    if not (tail_str.isdigit() and len(tail_str) == 7):
        return unit_accession
    tail = int(tail_str)
    if onto == _DEFAULT_UNIT_ONTOLOGY:
        return tail
    return [onto, tail]


def decode_tail(tail: int, ontology: str = _DEFAULT_PARAM_ONTOLOGY) -> str:
    """Reconstruct an accession string from a tail integer and ontology prefix."""
    return f"{ontology}:{tail:07d}"


def _is_numeric_tail(value: object) -> bool:
    """True for a wire numeric tail: a CBOR integer (never a boolean) in 0..9999999."""
    return type(value) is int and 0 <= value <= _MAX_NUMERIC_TAIL


def decode_unit_tail(tail: int | list | str) -> str:
    """Reconstruct a unit accession string from its wire form (int = UO: default,
    list = [ontology, tail], str = full accession).

    The wire form is checked, not the ontology: spectrl never resolves a release
    or asserts that a term exists. What is rejected here is a value that is not
    one of the three shapes section 3 defines -- a boolean read as a tail, a
    pair with a trailing extra member, a tail outside 0..9999999 (which would
    not survive seven-digit reconstruction), or a string that is not an
    accession. Without this the decoder emits accessions its own encoder
    refuses, and disagrees with the TypeScript reader about token validity.
    """
    if isinstance(tail, str):
        if not _ACCESSION_RE.fullmatch(tail):
            raise ValueError(f"invalid CV unit accession {tail!r}")
        return tail
    if isinstance(tail, list):
        if len(tail) != 2 or not isinstance(tail[0], str) or not _PREFIX_RE.fullmatch(tail[0]):
            raise ValueError("a CV unit ontology pair must be [prefix, tail]")
        if not _is_numeric_tail(tail[1]):
            raise ValueError(f"CV unit tail must be an integer in 0..{_MAX_NUMERIC_TAIL}")
        return f"{tail[0]}:{tail[1]:07d}"
    if not _is_numeric_tail(tail):
        raise ValueError(f"CV unit tail must be an integer in 0..{_MAX_NUMERIC_TAIL}")
    return f"{_DEFAULT_UNIT_ONTOLOGY}:{tail:07d}"


# Ion mobility array tails
ION_MOBILITY_ARRAY_TAILS: dict[str, int] = {decode_tail(tail): tail for tail in _ION_MOBILITY_TAILS}
